### DISCUSSION TITLE
Mesh with multiple roots

### DISCUSSION THREAD

**Discussion by rahav** [contiki-ng/contiki-ng#1636] [Q&A] [answered]
Is it possible in Contiki to have a mesh style network where each node can send a message to any root? For example, using the diagram we can have any of the nodes, say `node C`, **unicast** a message to `root 1` and another message to `root 2`. Alternatively a node could multicast the same message to a subset of the roots.

Similarly, can the roots **unicast** a message to to any of the nodes? No coordination among the roots is needed for this use case, it only requires that at any time, any of the roots, could send a message to any node, whether it is in its direct DODAG or not.

<img width="236" alt="Screen Shot 2021-09-07 at 09 47 29" src="https://user-images.githubusercontent.com/2912950/132365122-f46c6cb8-d8a5-4af7-a0c2-a6f26907b602.png">

**Comment by atiselsts** [comment, marked as answer]
With RPL classic it is possible to have multiple RPL DAGs and multiple RPL instances. So this setup is possible, for example, each root could source its own DAG, and the network nodes would be able to join both DAGs. Sending message to one root can then be done by addressing it its DAG ID, sending message to the other root: by addressing to the other root's DAG ID.

I'm not familiar with multicast in Contiki-NG so cannot answer about that.

**Reply by rahav** [reply]
Thanks, that's what I though but had hoped it is possible with RPL-lite or some other protocol that I wasn't aware of. I was even going to be happy if I could manage such communication patterns at the application layer if I had to.

**Reply by rahav** [reply]
Thanks again @atiselsts. Given your answer I wanted to follow Contiki API for RPL classic but couldn't find any. Is it documented and I just missed it?

I was expecting to see an API for a node to query the addresses of all roots, for example. Then join any subset of them. I didn't find any of this, but more surprisingly I didn't find any documentation similar to [RPL-lite](https://contiki-ng.readthedocs.io/en/master/_api/group__rpl-lite.html)

**Reply by atiselsts** [reply]
Hmm, I was not aware that RPL classic does not show up on readthedocs.io. Perhaps the header files need some maintenance. For now you can look directly at the .h files - the types and function prototypes are properly commented, the problem is just that these comments do not show up in the docs.

> I was expecting to see an API for a node to query the addresses of all roots, for example

The API works at the level of RPL instances and RPL DAGs. There isn't really a separate structure for a "root node", for example, so you'll have to look at a `struct rpl_dag` and use the DAG id as the address of the root node.

**Reply by rahav** [reply]
Thanks for your help a few days ago. After three futile days I need your help, again. I hope this is okay. As you suggested I tried to make sense by deciphering the "/* Public RPL functions. */" section in `rpl.h`. It lists the functions signatures but there's no in-code documentation. Some of these functions seems to not have an implementation, so I can't make out the functionality from code either. Even when the function is implemented, with new terminology (I'm new to RPL) it is difficult. I even read the RFC but it is long, long...

All I need is to know what function to call for a node to get the list of all roots. Then have that node call another API to connect to however many, 2 to n roots it wants. If these API do not exist I'll be happy to develop and contribute them but as written above, I don't even know where to start.

I presume, once a node is connected (say to 2 roots) I'll be able to use UDP to send a message from it to any of the roots. Similarly, I presume any of these root will be able to send a UDP message to any of the regular nodes. This is the key use case I'm trying to study, if I'm mistaken I'll appreciate if you can let me know before I do the work. Using the diagram above I will need nodes `A` - `E` to be able to send message to `Root 1` and `Root 2`. And either root be able to send a message to any of `A` - `E`.

If there's another way to achieve that I'll be happy to know before jumping. To change to RPL-classic I added `MAKE_ROUTING = MAKE_ROUTING_RPL_CLASSIC` to the Makefile.

Any way you can help? Or add to the discussion someone that can?

Thanks in advance,
Rahav

**Reply by rahav** [reply]
@joakimeriksson @nvt Can you help please? Is the scenario asked in my original post above supported by RPL-Classic?

Although it would be nice, multicast is not a hard requirement. As long as a node can find out the address of all roots, it can unicast to them one at a time.

I've been digging through code and after a week of trying out various approaches cannot get any node to get connected to, or even get the address of, more than one root.

**Reply by atiselsts** [reply]
> All I need is to know what function to call for a node to get the list of all roots. Then have that node call another API to connect to however many, 2 to n roots it wants. If these API do not exist I'll be happy to develop and contribute them but as written above, I don't even know where to start.

Joakim and Nicolas can add to this or correct me. From what I know, there is no such function because that's not how RPL works. The core abstractions in RPL are instances and DAGs. There is a global table of RPL instances, and each instance has a table of DAGs. In the normal case only one instance is used and one DAG is used, but in theory a node can join more than one of each. A DAG has `dag_id` member variable, which is an address of the DAG's root node. At least the Contiki's implementation, each DAG has a single root.

If you find functions that are declared but not implemented, by all means, raise a new Github issue describing the problem.

**Reply by rahav** [reply]
> but in theory a node can join more than one of each [instances, DAGs].

This is why I choose Contiki, it seemed to be a robust implementation of RPL RFC. However, I can't find a function that allows a node to query the available DAGs in an instance; and join any of these DAG.

Or alternatively, query the available instances and join DAGs from there (on top of what it is already connected to).

**Reply by atiselsts** [reply]
There is no option to query instances or DAGs ("query" as in sending messages to the network asking for this information). A node should wait for DIO messages from its neighbors in order to join a RPL instance's DAG. A node can only join a single DAG at the same time (but multiple instances). It's possible to speed up DIO message transmission by sending DIS messages to the neighbors. Does your reading of the RPL RFCs suggest there should be a functionality for active joining beyond this?

Also, let me stress again that multi-instance setups isn't really something that's typically used with Contiki. I guess that's why they were dropped in RPL Lite. There are no regression tests for multi-instance networks (there is one for a multi-DAG setup).

By looking at the code of RPL classic I see that there's still only a single RPL parent table (i.e. neighbor table) that is common for all instances. So I'm not sure if its actually possible to achieve what you want without extending the implementation.

**Reply by rahav** [reply]
Thank you. This helps.

**Reply by Blendinl** [reply]
@atiselsts @rahav Is it now possible to use multicasting to send a message from one node to multiple roots?
